src/GatewayProvider.ts

```diff
@@ -0,0 +1,80 @@
+import { Chain } from './types.js';
+import { withResolvers } from './utils.js';
+import {
+  JsonRpcPayload,
+  JsonRpcProvider,
+  JsonRpcResult,
+} from 'ethers/providers';
+import { FetchRequest } from 'ethers/utils';
+
+function shouldNeverBatch(payload: JsonRpcPayload) {
+  // see: LineaProver.ts:fetchProofs()
+  return payload.method === 'linea_getProof';
+}
+
+export class GatewayProvider extends JsonRpcProvider {
+  constructor(
+    fr: FetchRequest,
+    readonly chain: Chain
+  ) {
+    super(fr, chain, { staticNetwork: true });
+  }
+  override async _send(
+    payload: JsonRpcPayload | JsonRpcPayload[]
+  ): Promise<JsonRpcResult[]> {
+    if (Array.isArray(payload)) {
+      const ps: Promise<any>[] = [];
+      const batch: JsonRpcPayload[] = [];
+      const { promise, resolve, reject } = withResolvers<JsonRpcResult[]>();
+      const ret = payload.map((x) => {
+        if (shouldNeverBatch(x)) {
+          const p = this._sendWithRetry(x).then((x) => x[0]);
+          ps.push(p);
+          return p;
+        } else {
+          const i = batch.length;
+          batch.push(x);
+          return promise.then((v) => v[i]);
+        }
+      });
+      if (batch.length) {
+        ps.push(this._sendWithRetry(batch).then(resolve, reject));
+        await Promise.all(ps);
+      }
+      return Promise.all(ret);
+    }
+    return this._sendWithRetry(payload);
+  }
+  async _sendWithRetry(
+    payload: JsonRpcPayload | JsonRpcPayload[]
+  ): Promise<JsonRpcResult[]> {
+    let backoff = 250;
+    for (let attempt = 0; ; attempt++) {
+      try {
+        // ethers bug: this type is wrong
+        const results: any[] = await super._send(payload);
+        if (!results.some((x) => x.error?.code === 429)) {
+          // what the fuck is this
```

Comment:
XD Clean this up.

But also curious what you are so incensed about? XD.

Reply:
Oh. Was this what you were saying about a 200 status code with 429 specified in the response JSON?

Reply:
it's a bad implementation
it's an HTTP RPC (eg. single session), the docs say this only happens over WebSocket
instead causing HTTP 429 error, it does HTTP 200, and then does `{code: 429}`
for batched calls, say 100, you get 100 `{code: 429}`'s
by the time you've broken a batched response into its pieces, you've lost the header information
the headers might of contained `"retry-after"`, etc. but don't

Reply:
eurgh